File: Toggly.FeatureManagement.Flutter/toggly/lib/src/services/cache_lru.dart
import 'dart:convert';

/// Pure LRU index helpers matching `@ops-ai/toggly-hooks-types` `cache-lru`.

/// One tracked cache key's last-access metadata.
class CacheLruEntry {
  final int lastAccessed;

  const CacheLruEntry({required this.lastAccessed});

  Map<String, dynamic> toJson() => {'lastAccessed': lastAccessed};

  factory CacheLruEntry.fromJson(Map<String, dynamic> json) => CacheLruEntry(
        lastAccessed: (json['lastAccessed'] as num).toInt(),
      );
}

/// Sidecar LRU index persisted by cache providers via
/// `readCacheLruIndex` / `writeCacheLruIndex`.
class CacheLruIndex {
  final Map<String, CacheLruEntry> entries;

  const CacheLruIndex({required this.entries});

  Map<String, dynamic> toJson() => {
        'entries': entries.map((key, value) => MapEntry(key, value.toJson())),
      };
}

/// Empty index with no tracked keys.
CacheLruIndex emptyCacheLruIndex() => const CacheLruIndex(entries: {});

/// Parses sidecar JSON; corrupt or missing payloads become an empty index.
CacheLruIndex parseCacheLruIndex(String? raw) {
  if (raw == null || raw.isEmpty) {
    return emptyCacheLruIndex();
  }

  try {
    final decoded = jsonDecode(raw);
    if (decoded is! Map) {
      return emptyCacheLruIndex();
    }
    final entriesRaw = decoded['entries'];
    if (entriesRaw is! Map) {
      return emptyCacheLruIndex();
    }

    final entries = <String, CacheLruEntry>{};
    entriesRaw.forEach((key, value) {
      if (key is! String || value is! Map) {
        return;
      }
      final lastAccessed = value['lastAccessed'];
      if (lastAccessed is num) {
        entries[key] = CacheLruEntry(lastAccessed: lastAccessed.toInt());
      }
    });
    return CacheLruIndex(entries: entries);
  } catch (_) {
    return emptyCacheLruIndex();
  }
}

/// Serializes [index] to the JS-compatible JSON shape.
String serializeCacheLruIndex(CacheLruIndex index) =>
    jsonEncode(index.toJson());

/// Upserts [key] with [now] (epoch ms) as lastAccessed.
CacheLruIndex touchCacheLruKey(
  CacheLruIndex index,
  String key, {
  int? now,
}) {
  final timestamp = now ?? DateTime.now().millisecondsSinceEpoch;
  return CacheLruIndex(
    entries: {
      ...index.entries,
      key: CacheLruEntry(lastAccessed: timestamp),
    },
  );
}

/// Drops [keys] from the index.
CacheLruIndex removeCacheLruKeys(CacheLruIndex index, List<String> keys) {
  final entries = Map<String, CacheLruEntry>.from(index.entries);
  for (final key in keys) {
    entries.remove(key);
  }
  return CacheLruIndex(entries: entries);
}

/// Oldest keys to remove so [index] length is at most [maxKeys].
///
/// Skips [protectKey] / [protectKeys] (typically the key(s) just written for
/// the same evaluation context — e.g. flags + variants siblings).
List<String> selectCacheLruKeysToEvict(
  CacheLruIndex index,
  int maxKeys, {
  String? protectKey,
  Iterable<String>? protectKeys,
}) {
  if (maxKeys <= 0) {
    return const [];
  }

  final keys = index.entries.keys.toList();
  final over = keys.length - maxKeys;
  if (over <= 0) {
    return const [];
  }

  final protected = <String>{
    if (protectKey != null) protectKey,
    ...?protectKeys,
  };

  keys.sort((a, b) {
    final byAccess = index.entries[a]!.lastAccessed
        .compareTo(index.entries[b]!.lastAccessed);
    if (byAccess != 0) {
      return byAccess;
    }
    return a.compareTo(b);
  });

  final toEvict = <String>[];
  for (final key in keys) {
    if (toEvict.length >= over) {
      break;
    }
    if (protected.contains(key)) {
      continue;
    }
    toEvict.add(key);
  }

  return toEvict;
}

/// True when a positive finite max is configured.
bool isCacheLruEnabled(int? maxCacheKeys) {
  return maxCacheKeys != null && maxCacheKeys > 0;
}

File: toggly-hooks-types/src/cache-lru.ts
export interface CacheLruEntry {
  lastAccessed: number;
}

export interface CacheLruIndex {
  entries: Record<string, CacheLruEntry>;
}

export function emptyCacheLruIndex(): CacheLruIndex {
  return { entries: {} };
}

export function parseCacheLruIndex(raw: string | null | undefined): CacheLruIndex {
  if (!raw) {
    return emptyCacheLruIndex();
  }

  try {
    const parsed = JSON.parse(raw) as CacheLruIndex;
    if (!parsed || typeof parsed !== 'object' || typeof parsed.entries !== 'object' || !parsed.entries) {
      return emptyCacheLruIndex();
    }

    const entries: Record<string, CacheLruEntry> = {};
    for (const [key, value] of Object.entries(parsed.entries)) {
      if (!value || typeof value !== 'object') {
        continue;
      }
      const lastAccessed = (value as CacheLruEntry).lastAccessed;
      if (typeof lastAccessed === 'number' && Number.isFinite(lastAccessed)) {
        entries[key] = { lastAccessed };
      }
    }
    return { entries };
  } catch {
    return emptyCacheLruIndex();
  }
}

export function serializeCacheLruIndex(index: CacheLruIndex): string {
  return JSON.stringify({ entries: index.entries });
}

export function touchCacheLruKey(
  index: CacheLruIndex,
  key: string,
  now: number = Date.now(),
): CacheLruIndex {
  return {
    entries: {
      ...index.entries,
      [key]: { lastAccessed: now },
    },
  };
}

export function removeCacheLruKeys(index: CacheLruIndex, keys: string[]): CacheLruIndex {
  const entries = { ...index.entries };
  for (const key of keys) {
    delete entries[key];
  }
  return { entries };
}

function protectedKeySet(options?: {
  protectKey?: string;
  protectKeys?: string[];
}): Set<string> {
  const keys = new Set<string>();
  if (options?.protectKey) {
    keys.add(options.protectKey);
  }
  if (options?.protectKeys) {
    for (const key of options.protectKeys) {
      if (key) {
        keys.add(key);
      }
    }
  }
  return keys;
}

/**
 * Oldest keys to remove so the index length is at most `maxKeys`.
 *
 * Skips keys in `protectKeys` / `protectKey` (typically the key(s) just written
 * for the same evaluation context — e.g. flags + variants siblings).
 */
export function selectCacheLruKeysToEvict(
  index: CacheLruIndex,
  maxKeys: number,
  options?: { protectKey?: string; protectKeys?: string[] },
): string[] {
  if (!Number.isFinite(maxKeys) || maxKeys <= 0) {
    return [];
  }

  const limit = Math.floor(maxKeys);
  if (limit <= 0) {
    return [];
  }

  const keys = Object.keys(index.entries);
  const over = keys.length - limit;
  if (over <= 0) {
    return [];
  }

  const protectedKeys = protectedKeySet(options);
  const sorted = keys
    .slice()
    .sort((a, b) => (index.entries[a].lastAccessed - index.entries[b].lastAccessed) || a.localeCompare(b));

  const toEvict: string[] = [];
  for (const key of sorted) {
    if (toEvict.length >= over) {
      break;
    }
    if (protectedKeys.has(key)) {
      continue;
    }
    toEvict.push(key);
  }

  return toEvict;
}

/** True when a positive finite max is configured. */
export function isCacheLruEnabled(maxCacheKeys: number | null | undefined): boolean {
  return typeof maxCacheKeys === 'number' && Number.isFinite(maxCacheKeys) && maxCacheKeys > 0;
}
